### ISSUE TITLE
Audit log does not differentiate between domain blocks and domain limits

### ISSUE THREAD

**Issue by rixx** [mastodon/mastodon#32442]
### Steps to reproduce the problem

1. Limit a domain
2. Then block a domain
3. Look at the audit log: Two entries saying "User X blocked domain Y"

### Expected behaviour

Two differnt audit log entries: Domain limited vs domain suspended

### Actual behaviour

The same audit log entry appears twice

### Detailed description

I think we all agree that the audit log is not a strict audit log in the original use of the term, but being able to figure out after the fact when a domain was limited vs blocked (especially as those are escalations in behaviour) would be kinda nice.

If that's not possible at all, it would still be better if the entry just read "domain block was modified", in order to make it explicitly clear that you can't know what happened (rather than being confused at seeing the same entry twice).	

### Mastodon instance

chaos.social

### Mastodon version

v4.3.0

### Browser name and version

Browser-independent

### Operating system

OS independent

**Comment by ThisIsMissEm** [comment]
@mjankowski I know you're currently working on the audit logging code, I'm not sure if it'd be possible to vary the log line based on the severity of the DomainBlock?

**Comment by mjankowski** [comment]
I think that at the level of displaying the audit logs in that view, its all coming from i18n strings which just accept `name` and `target` interpolation values.

That particular one is:

```
create_domain_block_html: "%{name} blocked domain %{target}"
```

And in `DomainBlock`:

```
  def to_log_human_identifier
    domain
  end
```

A hacky solution here (to avoid changing ALL the i18n keys and model methods) would be to change that method to add the severity to the domain ... but then that portion would not be i18n'd.

Alternative would be to shift the i18n'ing responsibility into that model method, so that different classes which are audit loggable could have different interpolation values. Maybe on this one something like `"{name} created a domain block (#{severity}) on #{domain}"` but for other models it would still just have the two (or 1? or more?) values to i18n...

I'd have to look more at overall design/usage of the audit logs to have a strong opinion.

**Comment by rixx** [comment]
I do feel like "blocked domain x" is incorrect – from a user perspective – when the domain was limited rather than blocked. I understand that it's the same model on the backend side, but that's not really what a user is thinking about when they see "domain x was blocked".

Regardless, if you want to keep the strings as they are to avoid retranslation, could you add a prefix like "[block]" / "[suspend]" and have that also be i18n? It's just that there is currently *no* way for a user to differentiate what happened there at all.

… which leads me to my other question: is that information even stored? In your proposal of `"{name} created a domain block (#{severity}) on #{domain}"`, where would you get the `severity` information from?

**Comment by mjankowski** [comment]
Yeah, sorry, just to clarify -- I completely agree with your premise here which is that the action we are doing is more granular than the description is representing and ideally we can clarify that. I should have acknowledged that before getting into the weeds on implementation stuff.

That said -- the model is called "DomainBlock" and has a "severity" attribute (which is one of silence/suspend/noop on the backend - maps to the "limit", "suspend", "none" in the ui). So it would be easy enough to access the value and pass it into something -- the issue is more that EVERY single auditable model currently has the same interface into providing these strings, when probably they should have unique ones.

Let me contemplate more what the smallest possible change could be here.

**Comment by ThisIsMissEm** [comment]
Ah, yeah, because it only stores the model that changed: https://github.com/mastodon/mastodon/blob/main/app/models/admin/action_log_filter.rb

**Comment by mjankowski** [comment]
Right -- and then we run the account username, and whatever the result of the `log_target` helper is through the i18n.

The easiest thing to do to add *some* more information here but maybe not be most elegant, would be to add special casing for `DomainBlock` (and maybe others which have similar `enum`-style things?) to the helper method, and just cram in the `severity`, not run through i18n, as plain text there outside the link to the domain.

Maybe more elegant long term thing ... basically blow up that whole (long, nested!) helper method and give every auditable class a corresponding presenter class. Would need to know "is this linkable?" and then have plain text option when not, and link text plus link target when yes. Would allow all types to be basically infinitely flexible on string (as long as original source data still present) ... but would probably be quite the shuffle to the i18n strings.

Maybe I'll try that in a branch for 2-3 types and see how it feels.

**Comment by mjankowski** [comment]
Something which occurs to me here is that we don't actually have access to the historical information about what the state change was in the audit log. We know the time it happened, who did it, the relevant "target" (domain block in this example) ... and then we do have some "meta" information fields but those are used basically for capturing information about how to link to the thing.

There is a "human information" value, which in theory could have a plain text string with as much info as desired, but in practice has basically a single data value (domain, ip address, "acct" string, etc - depends on type of thing).

This also means that even if we added a feature or change to how the string is created, we don't actually have access to the historical record of what the escalation/de-escalation in severity was, in this example. We could add it for future changes, but not inspect past changes.

The way this works on other models - for appeal for example, is that the action that gets logged is not just a generic create/update/destroy (like on domain blocks), but something like "approve" or "reject", for example ... even though those are both technically "updates", that action logging lets us pull out the correct description later.

Changing the way domain blocks are logged to have more explicit actions that include the severity either as the action itself or some implied way might be useful here.

I found a few refactor ideas while looking at this.

**Comment by rixx** [comment]
> Something which occurs to me here is that we don't actually have access to the historical information about what the state change was in the audit log.

Yeah, that's why I asked where you'd get the information from. Following my confusion, I took to the database today after I saw your suggestion, and was a bit surprised to find that … Mastodon doesn't store this information (or rather, I had hoped to overlook something). Thank you for looking into it!
